sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/dofn/ReadChangeStreamPartitionDoFn.java

```diff
@@ -146,6 +156,17 @@ public TimestampRange initialRestriction(@Element PartitionMetadata partition) {
     return TimestampRange.of(startTimestamp, endTimestamp);
   }
 
+  @GetSize
+  public double getSize(@Element PartitionMetadata partition, @Restriction TimestampRange range)
+      throws Exception {
+    final BigDecimal timeGapInSeconds =
+        BigDecimal.valueOf(newTracker(partition, range).getProgress().getWorkRemaining());
```

Comment:
Instead of creating a new tracker here you can just receive a parameter of type `RestrictionTracker` (as per https://github.com/apache/beam/blob/bff4be488e12415cedc2e713f4e5129a18a14497/sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFn.java#L965-L967)

Reply:
Done.

Reply:
Actually, I can't make this change since I got the following error: 

```
/Users/hengfeng/src/github/beam/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/dofn/ReadChangeStreamPartitionDoFn.java:163: error: cannot find symbol
    final BigDecimal timeGapInSeconds = BigDecimal.valueOf(tracker.getProgress().getWorkRemaining());
                                                                  ^
  symbol:   method getProgress()
  location: variable tracker of type RestrictionTracker<TimestampRange,Timestamp>
```

`RestrictionTracker` interface does not have a `getProgress` method which comes from `hasProgress` interface.

Reply:
Ah ok. Do you think we could extract the progress tracking into its own class that is used both here and in the restriction tracker? You would be able to create a single instance in the setup method and re-use it? I feel it might be overkill to create a new restriction tracker inside this method.

Reply:
```
Do you think we could extract the progress tracking into its own class that is used both here and in the restriction tracker? You would be able to create a single instance in the setup method and re-use it?
```
I don't think we should do this. If you read the comment in https://github.com/apache/beam/blob/bff4be488e12415cedc2e713f4e5129a18a14497/sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFn.java#L965-L967, it also creates a new tracker and initializes it. It is actually the same. 

Also, KafkaIO creates a new restriction tracker as well: https://github.com/apache/beam/blob/8af69d7d4fe73fc94658f7d5b3d6721e969c7b1c/sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/ReadFromKafkaDoFn.java#L280

Reply:
I found this way: 

https://github.com/apache/beam/blob/8af69d7d4fe73fc94658f7d5b3d6721e969c7b1c/sdks/java/fn-execution/src/main/java/org/apache/beam/sdk/fn/splittabledofn/RestrictionTrackers.java#L105

Basically, cast a restriction tracker to a hasProgress interface. I updated it accordingly.

Reply:
Unfortunately, I define the getSize as follows: 

```
  @GetSize
  public double getSize(
      @Element PartitionMetadata partition,
      @Restriction TimestampRange range,
      RestrictionTracker<TimestampRange, com.google.cloud.Timestamp> tracker)
```

But the compile is okay, but I got the following error when running a pipeline: 

```
java.lang.IllegalArgumentException: org.apache.beam.sdk.io.gcp.spanner.changestreams.dofn.ReadChangeStreamPartitionDoFn, @GetSize getSize(PartitionMetadata, TimestampRange, RestrictionTracker): Illegal p
arameter type: RestrictionTrackerParameter{trackerT=org.apache.beam.sdk.transforms.splittabledofn.RestrictionTracker<org.apache.beam.sdk.io.gcp.spanner.changestreams.restriction.TimestampRange, com.googl
e.cloud.Timestamp>}
    at org.apache.beam.sdk.transforms.reflect.DoFnSignatures$ErrorReporter.throwIllegalArgument (DoFnSignatures.java:2397)
    at org.apache.beam.sdk.transforms.reflect.DoFnSignatures.checkParameterOneOf (DoFnSignatures.java:1300)
    at org.apache.beam.sdk.transforms.reflect.DoFnSignatures.analyzeGetSizeMethod (DoFnSignatures.java:2230)
    at org.apache.beam.sdk.transforms.reflect.DoFnSignatures.parseSignature (DoFnSignatures.java:757)
    at org.apache.beam.sdk.transforms.reflect.DoFnSignatures.lambda$getSignature$0 (DoFnSignatures.java:294)
    at java.util.HashMap.computeIfAbsent (HashMap.java:1220)
    at org.apache.beam.sdk.transforms.reflect.DoFnSignatures.getSignature (DoFnSignatures.java:294)
    at org.apache.beam.sdk.transforms.ParDo.validate (ParDo.java:614)
    at org.apache.beam.sdk.transforms.ParDo.of (ParDo.java:403)
    at org.apache.beam.sdk.io.gcp.spanner.SpannerIO$ReadChangeStream.expand (SpannerIO.java:1628)
    at org.apache.beam.sdk.io.gcp.spanner.SpannerIO$ReadChangeStream.expand (SpannerIO.java:1368)
    at org.apache.beam.sdk.Pipeline.applyInternal (Pipeline.java:548)
    at org.apache.beam.sdk.Pipeline.applyTransform (Pipeline.java:482)
    at org.apache.beam.sdk.values.PBegin.apply (PBegin.java:44)
    at org.apache.beam.sdk.Pipeline.apply (Pipeline.java:177)
    at com.examples.spanner.streaming.SpannerChangeStreams.main (SpannerChangeStreams.java:74)
    at org.codehaus.mojo.exec.ExecJavaMojo$1.run (ExecJavaMojo.java:254)
    at java.lang.Thread.run (Thread.java:833)
```

Reply:
Revert it back. Also, I found another example in the existing codebase: https://github.com/apache/beam/blob/8af69d7d4fe73fc94658f7d5b3d6721e969c7b1c/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsublite/internal/PerSubscriptionPartitionSdf.java#L124-L132